csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -755,588 +755,499 @@ module ControlFlow {
+          or
+          rec = TLastRecLoopBodyAbnormal() and
+          not c0 instanceof NormalCompletion and
+          not c0 instanceof ContinueCompletion and
+          not c0 instanceof BreakCompletion and
+          c = c0
+        )
+        or
+        // Last `catch` clause inherits throw completions from the `try` block,
+        // when the clause does not match
+        exists(SpecificCatchClause scc |
+          scc = cfe and
+          scc.isLast() and
+          throwMayBeUncaught(scc, c)
+        |
+          // Incompatible exception type: clause itself
+          result = scc
+          or
+          // Incompatible filter
+          result = lastSpecificCatchClauseFilterClause(scc)
+        )
+        or
+        cfe = any(TryStmt ts |
+            result = getBlockOrCatchFinallyPred(ts, c) and
+            (
+              // If there is no `finally` block, last elements are from the body, from
+              // the blocks of one of the `catch` clauses, or from the last `catch` clause
+              not ts.hasFinally()
+              or
+              // Exit completions ignore the `finally` block
+              c instanceof ExitCompletion
+            )
+            or
+            // If the `finally` block completes normally, it resumes any non-normal
+            // completion that was current before the `finally` block was entered
+            exists(NormalCompletion c0 |
+              result = lastTryStmtFinally(ts, c0) and
+              (
+                exists(getBlockOrCatchFinallyPred(ts, any(NormalCompletion nc))) and
+                c = c0
+                or
+                exists(getBlockOrCatchFinallyPred(ts, c)) and
+                not c instanceof NormalCompletion
+              )
+            )
+          )
       }
 
       pragma[nomagic]
-      private ControlFlowElement lastConditionalQualifiableExprChildExpr(
-        ConditionallyQualifiedExpr cqe, int i, Completion c
+      private ControlFlowElement lastRec(
+        ControlFlowElement cfe, Completion c, TLastRecComputation rec
       ) {
-        result = last(cqe.getChildExpr(i), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastObjectCreationArgument(ObjectCreation oc, int i, Completion c) {
-        result = last(getObjectCreationArgument(oc, i), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastObjectCreationInitializer(ObjectCreation oc, Completion c) {
-        result = last(oc.getInitializer(), c)
+        result = last(lastNonRec(cfe, TRec(rec)), c)
       }
 
       pragma[nomagic]
-      private ControlFlowElement lastArrayCreationInitializer(ArrayCreation ac, Completion c) {
-        result = last(ac.getInitializer(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastArrayCreationLengthArgument(
-        ArrayCreation ac, int i, Completion c
+      private ControlFlowElement lastRecSpecific(
+        ControlFlowElement cfe, Completion c1, Completion c2
```

Comment:
Drop a parameter? This is only ever called with `c1 = c2`

Reply:
That is by choice, to prevent too eager joining on `c1 = c2`.